src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+import random
+import re
+import string
+from itertools import product
+
+import numpy as np
+import xarray
+from extra.data import PropertyNameError
+
+
+def mangle_device_id_underscore(device_id):
+    """Converts deviceId to property notation (underscore version)."""
+    return ''.join(c if c.isalnum() else '_' for c in device_id)
+
+
+def mangle_device_id_camelcase(device_id):
+    """Converts deviceId to property notation (camelcase version)."""
+    # Replace non alpha-numeric chars with blanks
+    mangled_id = ''.join(c if c.isalnum() else ' ' for c in device_id)
+    # Make camelCase
+    mangled_id = ''.join(mangled_id.title().split())
+    if mangled_id:
+        mangled_id = mangled_id[0].lower() + mangled_id[1:]
+    return mangled_id
+
+
+def mangle_pattern(pattern, mangle, **args):
+    """Applies notation conversion to the format pattern."""
+    mangled = mangle(pattern.format(**args))
+    for field, placeholder in args.items():
+        mangled = mangled.replace(mangle(placeholder), "{" + field + "}")
+    return mangled
+
+
+def guess_device_id(mangled_id, underscores=2):
+    """Tries to guess original id from mangled id."""
+    device_id = ''
+    for c in mangled_id:
+        u = c.upper()
+        if underscores > 0 and c == u:
+            device_id += '_'
+            underscores -= 1
+        device_id += u
+    return device_id
+
+
+class DetectorMotors:
+    """Base class to access motors of detectors movable groups."""
+    _position_key = "actualPosition"
+
+    def __init__(self, dc, detector_id, src_ptrn, key_ptrn, **dims):
+        """
+        Args:
+            dc (extra_data.DataCollection):
+                The data
+            detector_id (str):
+                The detector ID, e.g. SPB_IRU_AGIPD1M or SPB_IRU_AGIPD1M
+            src_ptrn (str):
+                Format string pattern for motor sources
+            key_ptrn (str):
+                Format string pattern for motor position keys
+            **dims (lists):
+                Agruments to expand patterns and generate all motor data keys
+        Raises:
+            SourceNameError:
+                If any motor source is not found
+            PropertyNameError:
+                If any motor position key is not found
+        """
+        self.detector_id = detector_id
+        self.src_ptrn = src_ptrn
+        self.key_ptrn = key_ptrn
+        self.dims = dims
+
+        names, coordinates = zip(*self.dims.items())
+        self.motor_ids = [(
+                src_ptrn.format(**dict(zip(names, values))),
+                key_ptrn.format(**dict(zip(names, values)))
+            ) for values in product(*coordinates)]
+        self.motor_labels = [
+            ''.join(f"{n.upper()}{v}" for n, v in zip(names, values))
+            for values in product(*coordinates)]
+
+        self.shape = tuple(len(c) for c in coordinates)
+        self.num_sources = np.prod(self.shape)
+        self.dc = dc.select(self.motor_ids)
```

Comment:
```suggestion
        self.dc = dc.select(self.motor_ids, require_all=True)
```

In general this shouldn't make a difference for control sources, but with multiple data aggregators writing a run it's possible that some of them start or finish a train before the others.

Reply:
Good point, thanks!